File: model/src/test/java/com/github/fburato/highwheelmodules/model/modules/AnonymousModuleTest.java
package com.github.fburato.highwheelmodules.model.modules;

import com.github.fburato.highwheelmodules.model.bytecode.ElementName;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.DynamicTest.dynamicTest;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@DisplayName("AnonymousModule")
public class AnonymousModuleTest {
    @FunctionalInterface
    public interface ModuleBuilder<T extends MatchingModule> {
        Optional<T> make(String... globs);
    }

    public static <T extends MatchingModule> List<DynamicTest> containsTests(ModuleBuilder<T> builder) {
        final String GLOB = "org.pitest.foo*";
        final T testee = builder.make(GLOB).get();
        return Arrays.asList(
                dynamicTest("should return true if element name matches pattern",
                        () -> assertThat(testee.contains(new ElementName("org.pitest.foo.Something"))).isTrue()),
                dynamicTest("should return true when multiple matchers are available", () -> {
                    final MatchingModule otherTestee = builder.make("a*", "b*").get();
                    assertThat(otherTestee.contains(new ElementName("afoo"))).isTrue();
                    assertThat(otherTestee.contains(new ElementName("bfoo"))).isTrue();
                }),
                dynamicTest("should return false if element name does not match pattern",
                        () -> assertThat(testee.contains(new ElementName("not.pitest.foo"))).isFalse()),
                dynamicTest("should return false if the element name does not match any pattern", () -> {
                    final MatchingModule otherTestee = builder.make("a*", "b*").get();
                    assertThat(otherTestee.contains(new ElementName("cfoo"))).isFalse();
                }));
    }

    public static <T extends MatchingModule> List<DynamicTest> makeTests(ModuleBuilder<T> builder) {
        return Arrays.asList(
                dynamicTest("should be empty if regex passed is invalid",
                        () -> assertThat(builder.make("[asdf")).isEmpty()),
                dynamicTest("should be empty if any regex passed is invalid",
                        () -> assertThat(builder.make("valid", "[asdf")).isEmpty()),
                dynamicTest("should be defined if regex passed is valid",
                        () -> assertThat(builder.make("org.pitest.*")).isNotEmpty()),
                dynamicTest("should be defined if all regexes passed are valid",
                        () -> assertThat(builder.make("org.pitest.*", "valid")).isNotEmpty()));
    }

    @Test
    @DisplayName("equals should work on the pattern literals")
    void testEquals() {
        assertThat(AnonymousModule.make("a", "b")).isEqualTo(AnonymousModule.make("a", "b"));
        assertThat(AnonymousModule.make("a")).isNotEqualTo(AnonymousModule.make("b"));
    }

    @Test
    @DisplayName("equals should work on the pattern literals in different order")
    void testEqualsOrder() {
        assertThat(AnonymousModule.make("a", "b")).isEqualTo(AnonymousModule.make("b", "a"));
    }

    @TestFactory
    @DisplayName("contains")
    public List<DynamicTest> containsAnonymous() {
        return containsTests(AnonymousModule::make);
    }

    @TestFactory
    @DisplayName("make")
    public List<DynamicTest> makeAnonymous() {
        return makeTests(AnonymousModule::make);
    }
}

File: model/src/test/scala/com/github/fburato/highwheelmodules/model/modules/AnonymousModuleTest.scala
package com.github.fburato.highwheelmodules.model.modules

import com.github.fburato.highwheelmodules.model.bytecode.ElementName
import org.mockito.scalatest.MockitoSugar
import org.scalatest.OneInstancePerTest
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

class AnonymousModuleTest
    extends AnyWordSpec
    with Matchers
    with MockitoSugar
    with OneInstancePerTest {

  "make" should {
    "be empty if regex passed is invalid" in {
      AnonymousModule.make(List("[asdf")) shouldBe None
    }

    "be empty if any regex passed is invalid" in {
      AnonymousModule.make(List("valid", "[asdf")) shouldBe None
    }

    "be defined if regex passed is valid" in {
      AnonymousModule.make(List("org.example.*")).isDefined shouldBe true
    }

    "be defined if all regexes passed are valid" in {
      AnonymousModule.make(List("org.example.*", "valid")).isDefined shouldBe true
    }
  }

  "contains" should {
    val globPattern = List("org.example.foo*", "org.example.bar*")
    val testee = AnonymousModule.make(globPattern).get

    "match if name matches any pattern" in {
      testee.contains(ElementName.fromString("org.example.foo.Something")) shouldBe true
      testee.contains(ElementName.fromString("org.example.bar.Something")) shouldBe true
    }

    "match with globs rather than regex" in {
      testee.contains(ElementName.fromString("org|example|foo|Something")) shouldBe false
    }

    "use all glob allowed characters" in {
      def makeTestee(glob: String): AnonymousModule =
        AnonymousModule.make(List(glob)).get

      def name(s: String): ElementName = ElementName.fromString(s)

      makeTestee("$a$").contains(name("$a$")) shouldBe true
      makeTestee("^a$").contains(name("a")) shouldBe false
      makeTestee("a?").contains(name("ab")) shouldBe true
      makeTestee("a?").contains(name("b")) shouldBe false
      makeTestee("a.").contains(name("ab")) shouldBe false
      makeTestee("a.").contains(name("a.")) shouldBe true
      makeTestee("a\\").contains(name("a\\")) shouldBe true
    }

    "not match if name doesn't match any pattern" in {
      testee.contains(ElementName.fromString("notorg.example.foo")) shouldBe false
      testee.contains(ElementName.fromString("org.example.zoo")) shouldBe false
      testee.contains(ElementName.fromString("org.zxample.foo")) shouldBe false
    }
  }

  "equals should work on pattern literals irrespective of order" in {
    AnonymousModule.make(List("a")).get shouldEqual AnonymousModule.make(List("a")).get
    AnonymousModule.make(List("a", "b")).get shouldEqual AnonymousModule.make(List("b", "a")).get
    AnonymousModule.make(List("a")).get should not equal AnonymousModule.make(List("a", "b")).get
    AnonymousModule.make(List("a", "b")).get should not equal AnonymousModule.make(List("a")).get
  }
}
